core/isolate.rs

```diff
@@ -17,13 +17,52 @@ use libc::c_void;
 use std::collections::VecDeque;
 use std::ffi::CStr;
 use std::ffi::CString;
+use std::fmt::{Debug, Formatter, Result as FmtResult};
 use std::sync::{Arc, Mutex, Once, ONCE_INIT};
 
 pub type Buf = Box<[u8]>;
-pub type Op = dyn Future<Item = Buf, Error = ()> + Send;
+
+pub trait Op: Future + Debug + Sized + Send {
+  fn new<F: Future<Item = Self::Item, Error = Self::Error> + Send + 'static>(
+    label: &'static str,
+    future: F,
+  ) -> Self;
+}
+
+pub struct BoxOp<I, E> {
+  pub label: &'static str,
+  pub future: Box<dyn Future<Item = I, Error = E> + Send + 'static>,
+}
+
+impl<I, E> Op for BoxOp<I, E> {
+  fn new<F: Future<Item = I, Error = E> + Send + 'static>(
```

Comment:
I would organize this differently. Just have the name be part of Behavior::dispatch's return value and move name reference to PendingOp.